src/main/kotlin/org/rust/ide/completion/AttributeCompletionProvider.kt

```diff
@@ -0,0 +1,164 @@
+        RustAttribute("unsafe_no_drop_flag", onStruct), // Feature gated as of 1.9
+        RustAttribute("doc", onAnyItem),
+        RustAttribute("rustc_on_unimplemented", onTrait), // Feature gated as of 1.9
+        RustAttribute("cfg", onFn),
+        RustAttribute("cfg_attr", onAnyItem),
+        RustAttribute("allow", onAnyItem),
+        RustAttribute("warn", onAnyItem),
+        RustAttribute("forbid", onAnyItem),
+        RustAttribute("deny", onAnyItem),
+        RustAttribute("lang", onFn),
+        RustAttribute("inline", onFn),
+        RustAttribute("derive", onStruct or onEnum))
+
+    override fun addCompletions(parameters: CompletionParameters,
+                                context: ProcessingContext?,
+                                result: CompletionResultSet) {
+
+        val suggestions = attributes.filter { it.appliesTo.accepts(parameters.position)}
+            .map { LookupElementBuilder.create(it.name) }
+        result.addAllElements(suggestions)
+    }
+
+    val elementPattern: ElementPattern<PsiElement> get() {
+        val outerAttrElem = psiElement(RustCompositeElementTypes.OUTER_ATTR)
+        val innerAttrElem = psiElement(RustCompositeElementTypes.INNER_ATTR)
+        val metaItemElem = psiElement(RustCompositeElementTypes.META_ITEM)
+            .andOr(psiElement().withParent(outerAttrElem), psiElement().withParent(innerAttrElem))
+        return psiElement().withParent(metaItemElem).withLanguage(RustLanguage);
+    }
+
+    private data class RustAttribute(val name: String, val appliesTo: ElementPattern<PsiElement>)
+
+    private val onStruct: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(3, RustStructItemElement::class.java)
+    }
+
+    private val onEnum: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(3, RustEnumItemElement::class.java)
+    }
+
+    private val onFn: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(3, RustFnItemElement::class.java)
+    }
+
+    private val onMod: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(3, RustModItemElement::class.java)
+    }
+
+    private val onStatic: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(3, RustStaticItemElement::class.java)
+    }
+
+    private val onStaticMut: ElementPattern<PsiElement> get() {
+        // TODO
+        return onStatic
+    }
+
+    private val onMacro: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(3, RustMacroItemElement::class.java)
+    }
+
+    private val onTupleStruct: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(2, psiElement().withChild(psiElement(RustStructTupleArgsElement::class.java)))
+    }
+
+    private val onCrate: ElementPattern<PsiElement> get() {
+        val inMain = virtualFile().withName("main.rs")
+        val inLib = virtualFile().withName("lib.rs")
```

Comment:
We don't have an instance of the psiElement to call the extension methods on.  `psiElement()` returns an instance of `PsiElementPattern.Capture<PsiElement>`.